File: node/src/events.ts
/**
 * Lifecycle events, emitted through an injectable sink.
 *
 * The default sink writes one JSON line to stdout — no dependencies. An Emitter
 * owns a sink, so a process hosting more than one service gives each its own
 * instead of sharing one global. `DEFAULT` is the process-wide fallback.
 *
 * `token.auth_failed` fires ONLY when a bearer credential was presented and
 * failed to resolve — never on the no-credential path, which runs on effectively
 * every request.
 */
export type EventSink = (event: string, fields: Record<string, unknown>) => void;

export const DEFAULT_SERVICE = "unknown";

export function stdoutSink(event: string, fields: Record<string, unknown>): void {
  process.stdout.write(`${JSON.stringify({ event, ...fields })}\n`);
}

export class Emitter {
  sink: EventSink;
  service?: string;

  constructor(opts: { sink?: EventSink; service?: string } = {}) {
    this.sink = opts.sink ?? stdoutSink;
    this.service = opts.service;
  }

  emit(event: string, fields: Record<string, unknown> = {}, service?: string): void {
    // A sink that throws must never break the request path: events are
    // observability, not control flow.
    try {
      this.sink(event, { service: service ?? this.service ?? DEFAULT_SERVICE, ...fields });
    } catch {
      /* never propagate */
    }
  }

  tokenCreated(service: string, handle: string, label: string, role: string, actor: string): void {
    this.emit("token.created", { handle, label, role, actor }, service);
  }

  tokenRevoked(service: string, handle: string, label: string, actor: string): void {
    this.emit("token.revoked", { handle, label, actor }, service);
  }

  tokenRelabelled(service: string, handle: string, oldLabel: string, newLabel: string, actor: string): void {
    this.emit("token.relabelled", { handle, old: oldLabel, new: newLabel, actor }, service);
  }

  tokenPolicyChanged(service: string, handle: string, field: string, oldValue: unknown, newValue: unknown, actor: string): void {
    this.emit("token.policy_changed", { handle, field, old: oldValue, new: newValue, actor }, service);
  }

  tokenImported(service: string, handle: string, label: string, role: string): void {
    this.emit("token.imported", { handle, label, role }, service);
  }

  tokenImportRejected(service: string, role: string, reason: string): void {
    this.emit("token.import_rejected", { role, reason }, service);
  }

  tokenAuthFailed(
    service: string,
    fields: { handle?: string | null; label?: string | null; digestPrefix?: string | null; reason: string },
  ): void {
    this.emit(
      "token.auth_failed",
      {
        handle: fields.handle ?? null,
        label: fields.label ?? null,
        digest_prefix: fields.digestPrefix ?? null,
        reason: fields.reason,
      },
      service,
    );
  }

  tokenFirstUse(service: string, handle: string, label: string): void {
    this.emit("token.first_use", { handle, label }, service);
  }
}

export const DEFAULT = new Emitter();

/** Reconfigure the process default's sink, or reset it to stdout. */
export function setSink(sink?: EventSink | null): void {
  DEFAULT.sink = sink ?? stdoutSink;
}

File: simsys_tokens/events.py
"""Lifecycle events, emitted through an injectable sink.

The sink is a callable ``(event: str, fields: dict) -> None``. The default writes
one JSON line to stdout, using the standard library only — so the package has no
required dependencies. An adopter who wants structured logging or metrics wires
their own sink, or uses one of the adapters in
:mod:`simsys_tokens.integrations`.

``token.auth_failed`` fires ONLY when a bearer credential was presented and
failed to resolve. Never on the no-credential path: ``authenticate()`` runs on
effectively every request, so emitting on every ``None`` return would fire on
every anonymous hit to a health check or a scrape.
"""
from __future__ import annotations

import json
import sys

DEFAULT_SERVICE = "unknown"


def _stdout_sink(event: str, fields: dict) -> None:
    sys.stdout.write(json.dumps({"event": event, **fields}, default=str) + "\n")


# Process-wide sink. One install per process is the norm; if an app mounts the
# surface twice, the last sink wins. Tests replace this to capture events.
_sink = _stdout_sink


def set_sink(sink=None) -> None:
    """Install a sink callable, or reset to the stdout default when None."""
    global _sink
    _sink = sink or _stdout_sink


def emit(event: str, **fields) -> None:
    # A sink that raises must never break the request path: events are
    # observability, not control flow.
    try:
        _sink(event, fields)
    except Exception:
        pass


def token_created(service, handle, label, role, actor):
    emit("token.created", service=service, handle=handle, label=label, role=role, actor=actor)


def token_revoked(service, handle, label, actor):
    emit("token.revoked", service=service, handle=handle, label=label, actor=actor)


def token_relabelled(service, handle, old, new, actor):
    emit("token.relabelled", service=service, handle=handle, old=old, new=new, actor=actor)


def token_policy_changed(service, handle, field, old, new, actor):
    emit("token.policy_changed", service=service, handle=handle, field=field,
         old=old, new=new, actor=actor)


def token_imported(service, handle, label, role):
    emit("token.imported", service=service, handle=handle, label=label, role=role)


def token_import_rejected(service, role, reason):
    emit("token.import_rejected", service=service, role=role, reason=reason)


def token_auth_failed(service, *, handle=None, label=None, digest_prefix=None, reason):
    emit("token.auth_failed", service=service, handle=handle, label=label,
         digest_prefix=digest_prefix, reason=reason)


def token_first_use(service, handle, label):
    emit("token.first_use", service=service, handle=handle, label=label)
